builtin/logical/pki/storage_cache_issuers.go

```diff
@@ -0,0 +1,385 @@
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if !c.invalid {
+		return nil
+	}
+
+	c.idSet = make(map[issuerID]bool)
+	c.nameIDMap = make(map[string]issuerID)
+
+	// Clear the LRU; this is necessary as some entries/bundles might've been deleted.
+	c.entries.Purge()
+	c.bundles.Purge()
+	c.parsedBundles.Purge()
+	c.infoBundles.Purge()
+
+	// List all issuers which exist.
+	strList, err := s.List(ctx, issuerPrefix)
+	if err != nil {
+		return err
+	}
+
+	// Reset the issuer and name caches, populating the LRU.
+	for _, issuerIdStr := range strList {
+		issuerId := issuerID(issuerIdStr)
+
+		// Fetch the specified issuer; it might've been deleted since the
+		// list and thus returns an empty entry.
+		rawEntry, err := s.Get(ctx, issuerPrefix+issuerIdStr)
+		if err != nil {
+			return err
+		}
+		if rawEntry == nil {
+			continue
+		}
+
+		var entry issuerEntry
+		if err := rawEntry.DecodeJSON(&entry); err != nil {
+			return err
+		}
+
+		c.idSet[issuerId] = true
+		if len(entry.Name) > 0 {
+			c.nameIDMap[entry.Name] = issuerId
+		}
+
+		// Greedily add this entry to the LRU. We don't add bundles here.
+		c.entries.Add(issuerId, &entry)
+	}
+
+	c.invalid = false
+	return nil
+}
+
+func (c *issuerStorageCache) listIssuers(ctx context.Context, s logical.Storage) ([]issuerID, error) {
+	needUnlock := true
+	c.lock.RLock()
+	defer func() {
+		if needUnlock {
+			c.lock.RUnlock()
+		}
+	}()
+
+	if c.invalid {
+		// Release our read lock so we can race to grab a write lock.
+		c.lock.RUnlock()
+		needUnlock = false
+
+		if err := c.reloadOnInvalidation(ctx, s); err != nil {
+			return nil, err
+		}
+
+		// Now re-read-lock.
+		c.lock.RLock()
+		needUnlock = true
+	}
+
+	// Now we can service the request as expected.
+	result := make([]issuerID, 0, len(c.idSet))
+	for entry := range c.idSet {
+		result = append(result, entry)
+	}
+
+	return result, nil
+}
+
+func (c *issuerStorageCache) issuerWithID(ctx context.Context, s logical.Storage, id issuerID) (bool, error) {
+	needUnlock := true
+	c.lock.RLock()
+	defer func() {
+		if needUnlock {
+			c.lock.RUnlock()
+		}
+	}()
+
+	if c.invalid {
+		// Release our read lock so we can race to grab a write lock.
+		c.lock.RUnlock()
+		needUnlock = false
+
+		if err := c.reloadOnInvalidation(ctx, s); err != nil {
+			return false, err
+		}
+
+		// Now re-read-lock.
+		c.lock.RLock()
+		needUnlock = true
+	}
+
+	present, ok := c.idSet[id]
+	return ok && present, nil
+}
+
+func (c *issuerStorageCache) issuerWithName(ctx context.Context, s logical.Storage, name string) (issuerID, error) {
+	needUnlock := true
+	c.lock.RLock()
+	defer func() {
+		if needUnlock {
+			c.lock.RUnlock()
+		}
+	}()
+
+	if c.invalid {
+		// Release our read lock so we can race to grab a write lock.
+		c.lock.RUnlock()
+		needUnlock = false
+
+		if err := c.reloadOnInvalidation(ctx, s); err != nil {
+			return IssuerRefNotFound, err
+		}
+
+		// Now re-read-lock.
+		c.lock.RLock()
+		needUnlock = true
+	}
```

Comment:
This pattern repeats in all the getters.  Opportunity for a helper?

Reply:
Granted it'd need to pass the non-repeated part to the helper for proper defer semantics and without generics would break return values, so meh, maybe not.

Reply:
Yeah Steve has a fix-up commit he was playing with here: https://github.com/hashicorp/vault/commit/7c327497d2838e28e949020dafaed30c4d2d98e8 

I think there's bigger questions around locks though (there's the global `issuersLock` on the PKI `backend` directly too -- does it make sense to upgrade that to a read/write lock? What about write starvation w.r.t. lots of reads? Is the present double locking an issue? Are there races within this existing code? &c)

Overall I think this approach isn't viable in _this_ commit. (It was also my second time writing this code, the first attempt didn't pass any tests either. :D)

I think we'd probably be better served knowing that this is possible -- if there's issues resulting with slow performance, we _can_ fix them and we'd (even with fast-and-maybe-not-good code) get only a little better perf than <=1.10 (10%-ish). I think an actual design for this caching layer should perhaps be done a little more carefully. 

I'm on the fence about merging the struct changes, I think if we get consensus, we might want to do that anyways, just since there's a few nice refactors in it.

Perhaps I should just mark this one draft?